### ISSUE TITLE
IB implementation in DomainBed

### ISSUE THREAD

**Issue by alexrame** [ahujak/IB-IRM#1]
Thank you for the great and very interesting work.

I have a small question regarding the implementation of the IB penalty.
You stated in your paper (https://openreview.net/pdf?id=jlchsFOLfeF) that you "directly minimizes the entropy of the output layer". 
After a Gaussian approximation, this leads to minimizing $var(\Phi)$ (Section 4.1), _i.e._, minimizing the **logits variance** .

So I was surprised to see that you regularize **features variance** when you included your approach in the DomainBed benchmark.
https://github.com/facebookresearch/DomainBed/blob/2deb1502139ed1685fd3c0ad46ec78571d38c476/domainbed/algorithms.py#L1516.

Then I discovered this official repository https://github.com/ahujak/IB-IRM/blob/main/TerraIncognita/domainbed/algorithms.py, where the two versions are indeed implemented: IB_IRM_F_C targets features variance while IB_IRM_L_C targets logits variance.

So my question: does this mean that you used the features-based IB_IRM_F_C approach in Table 4 of your paper?
If so, why ? and have you tried to add a gaussian noise over the features, as recommended in Equation 1 from https://arxiv.org/pdf/2003.12537.pdf ?

Thanks in advance
Alexandre

**Comment by ahujak** [comment]
Hi

Thanks for appreciating our work!

Yes, for Table 4, we use  IB_ERM_F_C/IB_IRM_F_C. We mention and explain
this in the Supplement (
https://openreview.net/attachment?id=jlchsFOLfeF&name=supplementary_material)
page 49. "In both CMNIST datasets, we had experimented with placing the IB
penalty at the output layer (logits) and the penultimate layer (layer just
before the logits), and found that it is much more effective to place the
IB penalty on the penultimate layer".  Based on the experiments the general
sense that we have is that placing the penalty on the representation layer
is better for complex datasets such as the ones in Table 4. For linear
models, we do not need to separately create a representation layer and a
linear layer as that complicates things unnecessarily. For linear models
such as the ones in Table 3, we found that placing the penalty on the
logits works just fine. The theorems were presented for linear models so we
wanted to present the simplest possible penalty for that case. The theorems
can also be extended if you were to place a penalty on the representation
layer.

Best

**Comment by alexrame** [comment]
Thank you for the clear answer. That's now perfectly clear.

And have you tried to add some noise to the features ?
Theoretically, this is a reparameterization trick that enables having a distribution encoding of the input.
Practically, without noise, we can imagine a case in which the features get infinitely small but the classification weights get infinitely big ... and thus the bottleneck is not working anymore. Maybe the weight decay can prevent this, but still.

Best

**Comment by ahujak** [comment]
I did not realize there was a follow-up question. Sorry for the delay in reply. For some initial experiments, I recall trying noise addition at different points and that had not made much of a difference.
